File: CODE_ReFresh/JavaScript/Patterns/creational_abstract_factory/abstract_factory_1.js
class Sedan {
   constructor(tmName) {
      this.tmName = tmName;
      this.type = 'sedan';
   }
   create() {
      //..complex logic of creating Sedan
      console.log(`Create ${this.tmName}${this.type.toUpperCase()}`);
   }
}

class Coupe {
   constructor(tmName) {
      this.tmName = tmName;
      this.type = 'coupe';
   }
   create() {
      //..complex logic of creating Coupe
      console.log(`Create ${this.tmName}${this.type.toUpperCase()}`);
   };
}

/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */

class ToyotaCoupe extends Coupe {
   constructor() {
      super('Toyota');
   }
}

class ToyotaSedan extends Sedan {
   constructor() {
      super('Toyota');
   }
}

class FordCoupe extends Coupe {
   constructor() {
      super('Ford');
   }
}

class FordSedan extends Sedan {
   constructor() {
      super('Ford');
   }
}

/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */

class CarsFactory {
   createSedan(){
      throw new Error('CreateSedan() method is not defined');
   };
   createCoupe() {
      throw new Error('CreateSedan() method is not defined');
   };
}

/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */

class ToyotaFactory extends CarsFactory {
   createSedan() {
      return new ToyotaSedan();
   }
   createCoupe() {
      return new ToyotaCoupe();
   }
}

class FordFactory extends CarsFactory {
   createSedan() {
      return new FordSedan();
   }
   createCoupe() {
      return new FordCoupe();
   }
}

/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */

const factory = new ToyotaFactory();
const car = factory.createSedan(); // Creating instance of 'ToyotaSedan'
car.create();

File: CODE_Reminder/JavaScript/DesignPatterns/creational_abstract_factory/abstract_factory_1.java
public interface Sedan {}
public interface Coupe {}

public interface CarsFactory {
   Sedan createSedan();
   Coupe createCoupe();
}

/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */

public class ToyotaCoupe implements Coupe {
    public ToyotaCoupe() {
        System.out.println("Create ToyotaCoupe");
    }
}

public class ToyotaSedan implements Sedan {
    public ToyotaSedan() {
        System.out.println("Create ToyotaSedan");
    }
}

public class FordCoupe implements Coupe {
    public FordCoupe () {
        System.out.println("Create FordCoupe");
    }
}

public class FordSedan implements Sedan {
    public FordSedan() {
        System.out.println("Create FordSedan");
    }
}

/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */

public class ToyotaFactory implements CarsFactory {
   @Override
   public Sedan createSedan() {
      return new  ToyotaSedan();
   }

   @Override
   public Coupe createCoupe() {
      return new ToyotaCoupe();
   }
}

public class FordFactory implements CarsFactory {
   @Override
   public Sedan createSedan() {
      return new  FordSedan();
   }

   @Override
   public Coupe createCoupe() {
      return new FordCoupe();
   }
}

/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */

CarsFactory factory;

factory = new ToyotaFactory();

factory.createSedan();
